### ISSUE TITLE
Allowed Origins for `sql` endpoint

### ISSUE THREAD

**Issue by rotten** [MaterializeInc/materialize#13153]
### What version of Materialize are you using?

v0.26.2

### How did you install Materialize?

Docker image

### What is the issue?

I noticed that when I run materialize as a container locally, on my laptop, that the `/metrics`, `/memory`. and `/status` endpoints work fine.  However, when I deploy the container (in ECS), `/memory` throws a **401**.  The other endpoints work fine.

Digging into this a little I uncovered that it is actually the `/sql` endpoint, which is called by `/memory` that is throwing the **401**.

I also determined that, specifically, it is when the client sets the `Origin` header host value to anything other than localhost (or 127.0.0.1), it will throw the **401**.  The other endpoints don't seem to mind what the `Origin` header is set to.

I tried setting the environment variable `CORS_ALLOWED_ORIGIN` to various values, but it seems to be completely ignored.

While digging into this, I see this [line](https://github.com/MaterializeInc/materialize/blob/b2ecaa13b208be58bd1a74b8cb73e5d7fd7fc51e/src/environmentd/src/lib.rs#L405) in the code where the http::Server is getting an `allowed_origin` set.  The documentation for the hyper http server implies this should really be `allowed_origins` (with an **s**).  Is that correct?  I'm not a rust developer, so I could be completely wrong and maybe this is the way it is supposed to be.

I'm not sure why the server for `/sql` would behave differently than the other endpoints, `/metrics` and `/status`.  Other than overriding the `Origin` header in my client, it is not obvious how to get that endpoint to avoid throwing a 401.

**Comment by rotten** [comment]
Oh, leaving the `Origin` header completely out of the requests also gets around the **401** and works too.

**Comment by benesch** [comment]
Hi @rotten, thanks for the report! As you've discovered we shipped an improved security configuration with v0.26, but it has the downside that you now need to configure CORS appropriately for your environment. You'll need to set the CORS allowed origin to `PROTOCOL://HOST:PORT`, where each of those values is set based on what you, the user, will type into your browser to access Materialize. For example, if you were accessing Materialize via HTTP and HTTPS on the standard port at materialize.ecs.foocorp.com, you'd want to invoke `materialized` as:

```
materialized --cors-allowed-origin http://materialize.ecs.foocorp.com:6875 --cors-allowed-origin https://materialize.ecs.foocorp.com:6875
```

Could you give that a whirl and see if it works for you?

**Comment by benesch** [comment]
Forgot to mention: by default we allow CORS origins of localhost:

https://github.com/MaterializeInc/materialize/blob/10ff4ae61d835809fbc758450d519cd4b48929c4/src/materialized/src/bin/materialized/main.rs#L644-L655

That's probably why it works when you run locally but not within ECS.

**Comment by benesch** [comment]
So you're looking at the code on `main`, which has diverged substantially from v0.26.2. Here's the equivalent code in v0.26.2:

https://github.com/MaterializeInc/materialize/blob/10ff4ae61d835809fbc758450d519cd4b48929c4/src/materialized/src/bin/materialized/main.rs#L392-L393

Note the `MZ_` prefix, so `MZ_CORS_ALLOWED_ORIGIN`. 

But I would recommend avoiding the environment variable if possible. They're not robust against typos, and in this case there's no way to set two values for the parameter, if necessary, whereas you can just pass the CLI flag multiple times.

**Comment by benesch** [comment]
> I'll try the flags instead shortly. I'm still not sure the hyper server supports both allowed_origin and allowed_origins and fear that may be a typo.

Missed this part of your comment from earlier. It is not a typo! The code you linked originally refers to an `http::Server` type that is defined in our codebase, not one from an upstream library. It eventually results in a call to [`allow_origin`](https://docs.rs/tower-http/latest/tower_http/cors/struct.CorsLayer.html#method.allow_origin). Since Rust is a compiled language, any typo in here would result in a compile-time error.

**Comment by rotten** [comment]
Thanks, that makes sense.

Setting `MZ_CORS_ALLOWED_ORIGIN` works :-) ... to add a single domain.  I was not able to figure out how to get it to accept a list of origins (similar to the way MZ_LOG_FILTER will accept a comma separated list of filters).  Using the environment variable is preferred because we can use the same docker image and entry point script for all of our environments without having to build a loop into our dockerfile that constructs the command line.  I can get it to work, it is just that the environment variable approach would be simpler.

Anyway, after experimenting with various list configurations, I tried setting `MZ_CORS_ALLOWED_ORIGIN` to `"*"`.  That does not do what one might think.  It does not allow all origins.  Instead, I get a 401 for all origins unless I set my Origin header explicitly to `*` also.

I'm thinking this only really affects the `/sql` endpoint because it is called indirectly from the `/memory` endpoint.  The other endpoints, `/metrics` and `/status` don't call other endpoints.  So in our use case, we aren't setting Origin when we call those (at this time).

---

In short, thank you for your help working through this issue!  I think we are good to go now.

**Comment by rotten** [comment]
One other note:

`--cors-allowed-origin` does not show up in the list of options when you type `materialized --help`.

**Comment by benesch** [comment]
Fixed in #13190. Will be released in v0.26.3 later this week.
